OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs

```diff
@@ -0,0 +1,118 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2017 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Controls the spawning of actors. Attach this to the world actor.")]
+	public class ActorSpawnManagerInfo : ConditionalTraitInfo, Requires<MapCreepsInfo>
+	{
+		[Desc("Minimum number of actors")]
+		public readonly int Minimum = 0;
+
+		[Desc("Maximum number of actors")]
+		public readonly int Maximum = 4;
+
+		[Desc("Time (in ticks) between actor spawn.")]
+		public readonly int SpawnInterval = 6000;
+
+		[FieldLoader.Require]
+		[Desc("Name of the actor that will be spawned.")]
+		public readonly string[] Actors = { };
+
+		public readonly string ActorSignNotification = "";
+		public readonly string Owner = "Creeps";
+
+		[Desc("Type of ActorSpawner with which it connects.")]
+ 		public readonly HashSet<string> Types = new HashSet<string>() { "" };
+
+		public override object Create(ActorInitializer init) { return new ActorSpawnManager(init.Self, this); }
+	}
+
+	public class ActorSpawnManager : ConditionalTrait<ActorSpawnManagerInfo>, ITick, INotifyCreated
+	{
+		readonly ActorSpawnManagerInfo info;
+		readonly Lazy<TraitPair<ActorSpawner>[]> spawnPointActors;
+
+		bool enabled;
+		int spawnCountdown;
+		int actorsPresent;
+
+		public ActorSpawnManager(Actor self, ActorSpawnManagerInfo info) : base(info)
+		{
+			this.info = info;
+			spawnPointActors = Exts.Lazy(() => self.World.ActorsWithTrait<ActorSpawner>().Where(x => info.Types.Overlaps(x.Trait.Types)).ToArray());
+		}
+
+		void INotifyCreated.Created(Actor self)
+		{
+			enabled = self.Trait<MapCreeps>().Enabled;
+		}
+
+		void ITick.Tick(Actor self)
+		{
+			if (IsTraitDisabled)
+				return;
+
+			if (!enabled)
+				return;
+
+			if (!spawnPointActors.Value.Any())
+				return;
+
+			if (info.Maximum < 1 || actorsPresent >= info.Maximum)
+				return;
+
+			if (--spawnCountdown > 0 && actorsPresent >= info.Minimum)
```

Comment:
This might not ever happen but there could be an "underflow" here if the second condition never fires.

Reply:
There's no real case when it would happen, as adding actors is after this check

Reply:
I meant that --spawnCountdown might eventually underflow to a really big value after -2147483648 ticks... which will probably never happen since it will take way too long.

Reply:
@MunWolf Unless I'm miscalculating, we're talking about nearly 12000 hours at maximum game speed (50 ticks/sec).

I dare say whoever manages to set things up weirdly enough to trigger this, deserves the crash ;)